### PR TITLE
Registry functions: add :date and :time

### PR DESCRIPTION
**Pull request by ryzokuken**
Fix an error with `:dateTime`
Refactor parts of `:dateTime`
Add `:date`

### CODE REVIEW

File: spec.emu [hunk 1]
```diff
@@ -310,6 +310,71 @@ contributors: Eemeli Aro
...
+          _opts_: an ECMAScript value,
+        ): a Date object
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>It accepts an input that could be a Date object, a number or a string and converts them into Date objects.</dd>
+      </dl>
+
+        <emu-alg>
+          1. let value
```
Comment (eemeli) [inline spec.emu:326]: ?

File: spec.emu [hunk 2]
```diff
@@ -310,6 +310,71 @@ contributors: Eemeli Aro
...
+          1. If Type(_input_) is Object, then
+            1. If HasProperty(_input_, *"options"*) is *true*, then
+              1. Let _inputOptions_ be ? Get(_input_, *"options"*).
+              1. Perform ? Call(Object.assign, *undefined*, « _opts_, _inputOptions_ »).
```
Comment (eemeli) [inline spec.emu:330]: Option resolution should happen after value resolution, as it could fail and it has external visibility.

File: spec.emu [hunk 3]
```diff
@@ -310,6 +310,71 @@ contributors: Eemeli Aro
...
+            1. Let _valueOf_ be ? Get(_input_, *"valueOf"*).
+            1. If InstanceOfOperator(_input_, %Date%) is *false* and IsCallable(_valueOf_) is *true*, set _input_ to Call(_valueOf_, _input_).
```
Comment (eemeli) [inline spec.emu:332]: For instances of Date, we should not try to Get "valueOf".

File: spec.emu [hunk 4]
```diff
@@ -310,6 +310,71 @@ contributors: Eemeli Aro
...
+          1. If Type(_input_) is Number or String, let _value_ be ? Construct(%Date%, « _input_ »).
+          1. Else if Type(_input_) is Object, let _value_ be _input_.
```
Comment (eemeli) [inline spec.emu:334]: If the previous steps were to work on _value_ and not _input_, this line would be unnecessary.

File: spec.emu [hunk 5]
```diff
@@ -326,12 +391,15 @@ contributors: Eemeli Aro
         1. Let _number_ be CreateBuiltinFunction(_numberSteps_, *3*, *"number"*, « »).
         1. Let _stringSteps_ be the algorithm steps defined in <emu-xref href="#sec-messageformat-stringfunctions"></emu-xref>.
         1. Let _string_ be CreateBuiltinFunction(_stringSteps_, *3*, *"string"*, « »).
-        1. Let _dateTimeSteps_ be the algorithm steps defined in <emu-xref href="#sec-messageformat-stringfunctions"></emu-xref>.
-        1. Let _dateTime_ be CreateBuiltinFunction(_dateTimeSteps_, *3*, *"string"*, « »).
+        1. Let _dateTimeSteps_ be the algorithm steps defined in <emu-xref href="#sec-messageformat-datetimefunctions"></emu-xref>.
+        1. Let _dateTime_ be CreateBuiltinFunction(_dateTimeSteps_, *3*, *"dateTime"*, « »).
```
Comment (eemeli) [inline spec.emu:395]:
```suggestion
        1. Let _dateTime_ be CreateBuiltinFunction(_dateTimeSteps_, *3*, *"datetime"*, « »).
```

File: spec.emu [hunk 6]
```diff
@@ -525,35 +585,45 @@ contributors: Eemeli Aro
...
+          1. Let _value_ be ? HandleDateTimeInput(_input_, _opts_).
+          1. Let _dateStyle_ be ! Get(_options_, *"style"*).
+          1. If _dateStyle_ is *null* or *undefined*, then
+            1. Set _dateStyle_ to ! Get(_opts_, *"timeStyle"*).
+            1. If _dateStyle_ is *null* or *undefined*, then
+              1. Set _dateStyle_ to *"short"*.
+          1. For each String _optName_ of ? EnumerableOwnProperties(_options_, ~key~),
+            1. If _optName_ is not *"calendar"*, *"localeMatcher"*, *"hour12"*, *"hourCycle"*, *"numberingSystem"*, or *"timeZone"*, then
+              1. Let _optValue_ be ? Get(_options_, _optName_).
+              1. Perform ? Set(_opts_, _optName_, _optValue_, *"true"*).
```
Comment (eemeli) [inline spec.emu:620]: This is wrong; the only [supported option](https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.md#the-date-function) for `:date` is `style`.
```suggestion
          1. If _dateStyle_ is *null* or *undefined*, then
            1. Set _dateStyle_ to *"short"*.
```

**Review by eemeli — CHANGES_REQUESTED** [review]

File: spec.emu (hunk 1 above)
Comment (ryzokuken) [reply spec.emu:326]: Sorry for leaving that in

File: spec.emu (hunk 6 above)
Comment (eemeli) [reply spec.emu]: Ah, good point. But note that that code is not working on the `options` that are set by the expression options, but the `opt` which has had the `input.options` applied to it.

So while it's not allowing for any other expression options than `style`, it is supporting "locale options" when they're set on the operand's resolved value. This allows for e.g. this:

```js
const msg = 'Today is {$date :date style=long}'
const mf = new Intl.MessageFormat(msg, 'en')

const date = new Date()
mf.format({ date }) // 'Today is April 23, 2024'

date.options = { calendar: 'persian' }
mf.format({ date }) // 'Today is Ordibehesht 4, 1403 AP'
```

**Comment by ryzokuken** [comment]
@eemeli PTAL now. Regarding the first set of changes, I didn't entirely understand your comments (I think? I could be positively surprised) but I made some improvements that might address your concern.

In the second set of changes, your comment made me realize that I was misreading the code? Sorry for that, the updated spec should make a lot more sense.

**Review by eemeli — CHANGES_REQUESTED** [review]

File: spec.emu [hunk 7]
```diff
@@ -323,17 +323,15 @@ contributors: Eemeli Aro
...
-            1. If InstanceOfOperator(_input_, %Date%) is *false* and IsCallable(_valueOf_) is *true*, set _input_ to Call(_valueOf_, _input_).
-          1. If Type(_input_) is Number or String, let _value_ be ? Construct(%Date%, « _input_ »).
-          1. Else if Type(_input_) is Object, let _value_ be _input_.
-          1. If InstanceOfOperator(_value_, %Date%) is *false*, throw a *TypeError* exception.
-          1. Return _value_.
+            1. If InstanceOfOperator(_input_, %Date%) is *false*,then
+              1. If HasProperty(_input_, *"options"*) is *true*, then
+                1. Perform ? Call(Object.assign, *undefined*, « _opts_, Get(_input_, *"options"*) »).
+              1. Let _valueOf_ be ? Get(_input_, *"valueOf"*).
+              1. If IsCallable(_valueOf_) is *true*, set _input_ to Call(_valueOf_, _input_).
```
[the remark this answers is not in the document]
Comment (ryzokuken) [reply spec.emu:331]: I see, so basically if input is a non-date object, we're looking for `input.valueOf().options` and not `input.options`?

[the remark this answers is not in the document]
Comment (eemeli) [reply spec.emu:331]: No, still `input.options`. Note that I used _input_ and _value_ differently above than in your spec text. Looking at it now again, the corresponding `:number` code might also have some timing issues.

[the remark this answers is not in the document]
Comment (ryzokuken) [reply spec.emu]: I made an update, feel free to check them out whenever you can. In the meantime I'll check what `:numbers` does too. JFYI btw, now we've diverged from what the polyfill does (not that the spec needs to follow the polyfill, but just wanted to point that out): https://github.com/messageformat/messageformat/blob/main/packages/mf2-messageformat/src/functions/datetime.ts#L126-L146

[the remark this answers is not in the document]
Comment (eemeli) [reply spec.emu]: Very happy to update the polyfill to match the spec for a change, rather than vice versa

**Review by eemeli — APPROVED** [review]
Looks good, if we include the one refactor proposed inline.

**Comment by ryzokuken** [comment]
Overall looks good to me, although I still feel like we could avoid repeating the `if (input.options) {...}` part while keeping the rest as-is.

**Comment by ryzokuken** [comment]
@eemeli this might look better now :D

**Review by eemeli — APPROVED** [review]
Yup, looks good!

**Pull request merged**